File: frontend/src/types/risk.ts
/**
 * Risk Domain Types for Part 7C & Part 7D: Deterministic Risk & Overrides Engine
 * Represents base risk, adjusted risk, override triggers, feature vectors, and explainable contributions.
 */

export type RiskLevel = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export const RiskLevelEnum = {
  LOW: 'LOW' as const,
  MEDIUM: 'MEDIUM' as const,
  HIGH: 'HIGH' as const,
  CRITICAL: 'CRITICAL' as const,
};

export type RiskIndicator =
  | 'COMPLIANCE_DEFICIT'
  | 'RULE_FAILURES'
  | 'REVIEW_UNCERTAINTY'
  | 'PENDING_UNCERTAINTY'
  | 'MANDATORY_FAILURES'
  | 'INTEGRITY_CONCERNS';

export type RiskOverrideType =
  | 'RISK_FLOOR'
  | 'RISK_INCREMENT'
  | 'LEVEL_FLOOR'
  | 'SCORE_CAP'
  | 'REVIEW_ESCALATION';

export interface RiskContribution {
  indicator: RiskIndicator | string;
  name: string;
  raw_value: string;
  normalized_value: number | string;
  weight: number | string;
  weighted_contribution: number | string;
  reason: string;
}

export interface RiskOverride {
  rule_code?: string | null;
  override_type: RiskOverrideType | string;
  trigger: string;
  source_result_id?: string | null;
  source_requirement_id?: string | null;
  previous_score?: number | string | null;
  new_score?: number | string | null;
  previous_level?: string | null;
  new_level?: string | null;
  risk_floor?: number | string | null;
  risk_increment?: number | string | null;
  minimum_level?: string | null;
  reason: string;
  severity: 'INFO' | 'WARNING' | 'HIGH' | 'CRITICAL' | string;
}

export interface RiskFeatures {
  overall_compliance_score?: number | string | null;
  total_rules: number;
  applicable_rules: number;
  passed_count: number;
  fail_count: number;
  review_count: number;
  pending_count: number;
  not_applicable_count: number;

  mandatory_rules_count: number;
  mandatory_failure_count: number;
  critical_failure_count: number;

  integrity_rules_count: number;
  integrity_fail_count: number;
  integrity_review_count: number;
  cross_document_mismatch_count: number;
  low_confidence_count: number;

  scoring_complete: boolean;
  human_review_required: boolean;
}

export interface BidRiskAssessmentResponse {
  id?: string | null;
  bid_id: string;
  tender_id: string;
  risk_version: number;
  risk_formula_version: string;
  override_formula_version?: string;

  // Part 7C: Mathematical Base Risk
  base_risk_score?: number | string | null;
  base_risk_level?: RiskLevel | string | null;

  // Part 7D: Deterministic Adjusted Risk
  adjusted_risk_score?: number | string | null;
  adjusted_risk_level?: RiskLevel | string | null;

  // Overrides Applied
  override_applied?: boolean;
  override_count?: number;
  applied_overrides?: RiskOverride[];

  // Operational & Readiness Flags
  risk_complete: boolean;
  is_provisional: boolean;
  human_review_required: boolean;

  // Granular Signals & Explanations
  features: RiskFeatures;
  contributions: RiskContribution[];
  summary_reasons: string[];
  calculation_details?: Record<string, any>;
  calculated_at?: string | null;
}

File: backend/app/schemas/risk.py
"""
Risk Schemas for Part 7C & Part 7D: Deterministic Risk & Overrides Engine
Defines Pydantic request and response schemas for Risk APIs.
"""

from datetime import datetime
from decimal import Decimal
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class RiskContributionResponse(BaseModel):
    """Auditable itemized contribution of a single risk indicator."""
    model_config = ConfigDict(from_attributes=True)

    indicator: str
    name: str
    raw_value: str
    normalized_value: Decimal
    weight: Decimal
    weighted_contribution: Decimal
    reason: str


class RiskOverrideResponse(BaseModel):
    """Auditable record of an applied deterministic risk override or floor."""
    model_config = ConfigDict(from_attributes=True)

    rule_code: Optional[str] = None
    override_type: str
    trigger: str
    source_result_id: Optional[str] = None
    source_requirement_id: Optional[str] = None

    previous_score: Optional[Decimal] = None
    new_score: Optional[Decimal] = None

    previous_level: Optional[str] = None
    new_level: Optional[str] = None

    risk_floor: Optional[Decimal] = None
    risk_increment: Optional[Decimal] = None
    minimum_level: Optional[str] = None

    reason: str
    severity: str


class RiskFeaturesResponse(BaseModel):
    """Extracted feature vector metrics used in risk evaluation."""
    model_config = ConfigDict(from_attributes=True)

    overall_compliance_score: Optional[Decimal] = None
    total_rules: int = 0
    applicable_rules: int = 0
    passed_count: int = 0
    fail_count: int = 0
    review_count: int = 0
    pending_count: int = 0
    not_applicable_count: int = 0

    mandatory_rules_count: int = 0
    mandatory_failure_count: int = 0
    critical_failure_count: int = 0

    integrity_rules_count: int = 0
    integrity_fail_count: int = 0
    integrity_review_count: int = 0
    cross_document_mismatch_count: int = 0
    low_confidence_count: int = 0

    scoring_complete: bool = True
    human_review_required: bool = False


class BidRiskAssessmentResponse(BaseModel):
    """
    Complete risk assessment snapshot response containing:
    - Part 7C Base Risk (pure mathematical calculation)
    - Part 7D Adjusted Risk (post deterministic overrides & floors)
    - Completeness & Provisional flags
    - Feature vector & itemized indicator contributions
    - Applied override audit history and summary reasons
    """
    model_config = ConfigDict(from_attributes=True)

    id: Optional[str] = None
    bid_id: str
    tender_id: str
    risk_version: int
    risk_formula_version: str
    override_formula_version: str = "v1"

    # Part 7C: Mathematical Base Risk
    base_risk_score: Optional[Decimal] = None
    base_risk_level: Optional[str] = None

    # Part 7D: Deterministic Adjusted Risk
    adjusted_risk_score: Optional[Decimal] = None
    adjusted_risk_level: Optional[str] = None

    # Overrides Applied
    override_applied: bool = False
    override_count: int = 0
    applied_overrides: List[RiskOverrideResponse] = Field(default_factory=list)

    # Operational & Readiness Flags
    risk_complete: bool
    is_provisional: bool
    human_review_required: bool

    # Signals & Explanations
    features: RiskFeaturesResponse
    contributions: List[RiskContributionResponse] = Field(default_factory=list)
    summary_reasons: List[str] = Field(default_factory=list)
    calculation_details: Dict[str, Any] = Field(default_factory=dict)
    calculated_at: Optional[datetime] = None
